cli/azd/pkg/infra/provisioning/provider.go

```diff
@@ -21,7 +22,9 @@ type ProviderKind string
 type NewProviderFn func(ctx context.Context, env *environment.Environment, projectPath string, infraOptions Options) (Provider, error)
 
 var (
-	providers map[ProviderKind]NewProviderFn = make(map[ProviderKind]NewProviderFn)
+	//providersMutex guards providers. the lock should be taken before interacting with the map.
+	providersMutex sync.Mutex                     = sync.Mutex{}
```

Comment:
I ran into this in CI and so I would like to fix it quickly but I really don't love using the mutex in this way.  I wish there was some way that we didn't have to do this explicit registration.  I'm not sure if wire would end up helping us here.  If we are to have this long term, we should move away from it being a package global.

Reply:
I think I brought it up in the initial PR that I was wondering if we had a need to introduce a plugin model for statically linked code. I would have preferred just having "registration" be adding a switch case to a factory that constructs the appropriate provider since everything can be known statically.